brainstorming.md

```diff
@@ -0,0 +1,372 @@
+```
+
+Parameters:
+
+- ids - list of indices to read. Defaults to 'all'
+- column_names - the named columns to read and return. Defaults to all.
+- filter - [value filter](#value-filters) to apply (_TBD_). Defaults to no filter.
+
+> ⚠️ **Issue** - sdf.read() needs further definition for:
+>
+> - incremental/streaming queries - critical gap
+> - result ordering - would be useful, but also adds complexity (order by index? order by column value?). Most useful for streaming queries where the total result may be too large to load into memory and sort.
+> - filtering based upon values - need to write the section: [Value Filter](#value-filters)
+
+> ℹ️ **Note** -- with this definition for read(), it should be possible to implement the equivalent of the higher level functions in the current prototype:
+>
+> - ids() -- read(ids=all, column_names=['the index column name, eg, obs_id'])
+> - query()/attribute_filter() -- read(column_names=['ids', 'a_col'], filter='a_col == "foobar"')
+
+### Method: write()
+
+Write an Arrow.RecordBatch to the persistent object. As duplicate index values are not allowed, index values already present in the object are overwritten and new index values are added.
+
+```
+sdf.write(Arrow.RecordBatch values)          # write
+```
+
+Parameters:
+
+- values - an Arrow.RecordBatch containing all columns, including the index column. The schema for the values must match the schema for the SOMADataFrame.
+
+## SOMAMatrix
+
+> 🛑 **To be further specified** -- will look essentially the same as SOMADataFrame, extended with a second index dimension.
+
+In the following, `smx` is used to indicate an "instance" or "handle" to a SOMAMatrix.
+
+Summary:
+
+```
+SOMAMatrix.create(string uri, Arrow.Schema schema, string index_column_name) -> void
+
+SOMAMatrix.delete(string uri) -> void
+SOMAMatrix.exists(string uri) -> bool  # exists and is a SOMADataFrame
+SOMAMatrix.open(string uri, mode=read|write) -> smx
+smx.close()
+
+smx.shape -> (int, int)                     # length of each index dimension, always a list of length 2
+smx.ndims -> 2                              # returns a constant
+smx.schema -> Arrow.Schema                  # data schema, which must be the schema for an Arrow.RecordBatch
+smx.index_column_names -> (string, string)  # list of index column names, always a list of length 2
+
+smx.read(ids=[...]|all, column_names=[...]|all, filter=_TBD_) -> Arrow.RecordBatch
+smx.write(Arrow.RecordBatch values)          # write
+```
+
+## SOMA
+
+SOMA supports all SOMACollection functions, including access to the predefined fields (eg, `X`) using SOMACollection API. The primary difference in behavior between a SOMACollection and a SOMA is that the SOMA _should_ enforce naming and indexing conventions for the predefine fields (eg, `obs`), as define in the [SOMA data model](#soma).
+
+Summary:
+
+```
+SOMA.create(uri) -> soma
+SOMA.open(uri, mode=read|write) -> soma
```

Comment:
same as above -- i think open/close are storage-level concerns, not API-level concerns

Reply:
removed from latest